### PR TITLE
[KV Offload] Fix multi-node KV offloading state desynchronization and JAX dispatch…

### PR DESCRIPTION
**Pull request by amitkumar307d**
# Description

This PR fixes state desynchronization, JAX dispatch deadlocks, and resource leaks in Multi-Node KV Offloading, ensuring stability during high-concurrency benchmarks.

**Context & Problem:**
Previously, running high-concurrency benchmarks (like `prefix_repetition`) on large models (e.g., Qwen3-Coder 480B) with multi-node KV offloading enabled led to fatal crashes (`AssertionError`) and runtime hangs. 

1. **State Desynchronization:** In multi-node setups, completion signals (finished chunk IDs) from worker nodes were being discarded in the stats aggregation phase, causing the Scheduler to miss confirmations and desync buffer accounting.
2. **JAX Dispatch Deadlocks:** Running `jax.device_put` (D2H) in background threads caused non-deterministic dispatch orders across nodes, leading to low-level TPU runtime halts.
3. **Resource Leaks:** Duplicate hashes in a batch resulted in multiple physical CPU chunks being allocated for the same content, wasting CPU RAM.
4. **Data Corruption Races (RAW/WAW):** 
   - **Read-After-Write (RAW):** A fast host could trigger a Load on a slow host before the slow host finished its Save in the background.
   - **Write-After-Write (WAW):** Rapid slot recycling caused background threads to race when writing to the same CPU memory address, potentially overwriting newer data with older data.
5. **Thread Safety Gaps:** Concurrent modifications to the `OrderedDict` backing the `LocalCPUBackend` caused structural corruption under high load.

---

# Solution

- **Robust Distributed State Synchronization:** 
  - Implemented **`KVOffloadConnectorStats.aggregate`** using `collections.Counter` and the union (`|`) operator to correctly merge completion signals across all worker nodes without artificially multiplying chunk counts.
  - Fixed **`KVOffloadConnectorStats.num_finished_blocks`** to accurately count total finished chunks instead of just request entries.
- **Deterministic JAX Dispatch Order:** Moved **`jax.device_put`** (D2H dispatch) from background threads to the **main thread** in `TPUOffloadConnectorWorker.start_save_kv` (and batched version) to ensure globally aligned dispatch order across all TPU nodes.
- **Resource Leak Prevention:** Added **deduplication** of `chunk_hashes` in `OffloadManager.allocate_for_save` to ensure only one `CPUChunk` is allocated per unique content hash in a batch.
- **Resilient Bookkeeping Cleanup:** Relaxed strict assertions and replaced `remove()` with **`discard()`** in `update_connector_output` to handle redundant or late completion signals gracefully without crashing the engine.
- **Concurrency Hardening & Race Mitigation:**
  - **Local Read-Before-Write Sync:** Worker main thread tracks `_local_in_flight_saves` and blocks (`future.result()`) during `start_load_kv` if a Load is requested for a chunk still being saved locally.
  - **Dependency-Chained Serialization (WAW Fix):** Background saves targeting the exact same CPU slot are chained via future dependencies in the thread pool, enforcing strict FIFO order of writes without requiring global central barriers.
  - **Thread-Safe CPU Cache:** Secured `LocalCPUBackend` by wrapping critical sections (add, get, reclaim) with `threading.Lock` to protect the underlying dictionary from concurrent corruption.

---

# Tests

- Verified with **Qwen3-Coder 480B** on a multi-node TPU cluster running `prefix_repetition` benchmarks under high concurrency.
- Confirmed that the `AssertionError` state desync is resolved and the server runs to completion without JAX runtime hangs.

**Qwen3-Coder 480B - Server command:**
```bash
python3 -m vllm.entrypoints.openai.api_server \
--host=0.0.0.0 \
--port=8000 \
--tensor-parallel-size=16 \
--max-model-len=102400 \
--load-format=runai_streamer \
--kv-cache-dtype=fp8 \
--gpu-memory-utilization=0.5 \
--data-parallel-size=1 \
--max-num-batched-tokens=32768 \
--max-num-seqs=512 \
--model=Qwen/Qwen3-Coder-480B-A35B-Instruct \
--served-model-name=Qwen/Qwen3-Coder-480B-A35B-Instruct \
--enable-prefix-caching \
--async-scheduling \
--enable-expert-parallel \
--kv-transfer-config='{"kv_connector": "TPUOffloadConnector", "kv_connector_module_path": "tpu_inference.offload.tpu_offload_connector","kv_role": "kv_both", "kv_connector_extra_config": {"cpu_bytes_to_use": 107374182400, "lazy_offload": false}}'
```

**Qwen3-Coder-480B - Client Command:**
```bash
vllm bench serve \
  --backend=vllm \
  --model=Qwen/Qwen3-Coder-480B-A35B-Instruct \
  --dataset-name=prefix_repetition \
  --host=localhost \
  --port=8000 \
  --seed=123 \
  --num-prompts=256 \
  --max-concurrency=256 \
  --prefix-repetition-prefix-len=99000 \
  --prefix-repetition-suffix-len=1000 \
  --prefix-repetition-output-len=1024 \
  --prefix-repetition-num-prefixes=4 \
  --percentile-metrics='ttft,tpot,itl,e2el' \
  --ignore-eos
```

**Logs:**
```bash
(EngineCore pid=1772) INFO 07-01 14:57:30 [tpu_offload_connector.py:738] Request cmpl-ba5fb6d4de56c58b-0-b3e5550a: After accounting for 98816 computed tokens, reporting 0 tokens to load.
(EngineCore pid=1772) INFO 07-01 14:57:30 [tpu_offload_connector.py:672] Request cmpl-ae673998978b8581-0-b9fc8a5b: Found 98816 (out of 100000 existing tokens) matched tokens (193 blocks) in CPU backend (computed_blocks: 193, blocks_to_load: 0).
(EngineCore pid=1772) INFO 07-01 14:57:30 [tpu_offload_connector.py:738] Request cmpl-ae673998978b8581-0-b9fc8a5b: After accounting for 98816 computed tokens, reporting 0 tokens to load.
(APIServer pid=1214) INFO 07-01 14:57:31 [loggers.py:273] Engine 000: Avg prompt throughput: 3433.7 tokens/s, Avg generation throughput: 132.3 tokens/s, Running: 89 reqs, Waiting: 7 reqs, GPU KV cache usage: 96.3%, Prefix cache hit rate: 98.8%, External prefix cache hit rate: 0.0%
(APIServer pid=1214) INFO 07-01 14:57:31 [metrics.py:103] KV Transfer metrics: Num finished save chunks =51, Num finished load chunks=23
(APIServer pid=1214) INFO 07-01 14:57:41 [loggers.py:273] Engine 000: Avg prompt throughput: 1184.0 tokens/s, Avg generation throughput: 460.8 tokens/s, Running: 88 reqs, Waiting: 0 reqs, GPU KV cache usage: 94.6%, Prefix cache hit rate: 98.8%, External prefix cache hit rate: 0.0%
(APIServer pid=1214) INFO 07-01 14:57:41 [metrics.py:103] KV Transfer metrics: Num finished save chunks =28, Num finished load chunks=0
(APIServer pid=1214) INFO 07-01 14:57:51 [loggers.py:273] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 528.0 tokens/s, Running: 88 reqs, Waiting: 0 reqs, GPU KV cache usage: 94.6%, Prefix cache hit rate: 98.8%, External prefix cache hit rate: 0.0%
(APIServer pid=1214) INFO 07-01 14:58:01 [loggers.py:273] Engine 000: Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 519.9 tokens/s, Running: 85 reqs, Waiting: 0 reqs, GPU KV cache usage: 95.5%, Prefix cache hit rate: 98.8%, External prefix cache hit rate: 0.0%
```

**Results:**
```bash
============ Serving Benchmark Result ============
Successful requests:                     256       
Failed requests:                         0         
Maximum request concurrency:             256       
Benchmark duration (s):                  727.31    
Total input tokens:                      25600009  
Total generated tokens:                  262144    
Request throughput (req/s):              0.35      
Output token throughput (tok/s):         360.43    
Peak output token throughput (tok/s):    595.00    
Peak concurrent requests:                256.00    
Total token throughput (tok/s):          35558.83  
---------------Time to First Token----------------
Mean TTFT (ms):                          247942.56 
Median TTFT (ms):                        301235.40 
P99 TTFT (ms):                           595670.52 
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          261.68    
Median TPOT (ms):                        263.58    
P99 TPOT (ms):                           414.83    
---------------Inter-token Latency----------------
Mean ITL (ms):                           262.36    
Median ITL (ms):                         169.30    
P99 ITL (ms):                            2583.81   
----------------End-to-end Latency----------------
Mean E2EL (ms):                          515642.96 
Median E2EL (ms):                        584923.77 
P99 E2EL (ms):                           726457.76 
==================================================
```

# Checklist
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

**Comment by dannawang0221** [comment]
To enable KV cache offloading, please set TPU_OFFLOAD_NUM_CPU_CHUNKS and TPU_OFFLOAD_NUM_STAGING_BLOCKS. Note that the cpu_bytes_to_use parameter is ignored in tpu-inference.

Ref: [https://github.com/vllm-project/tpu-inference/blob/main/examples/offload/gke/benchmarks/deploy-cpu-offload.yaml#L98](https://github.com/vllm-project/tpu-inference/blob/main/examples/offload/gke/benchmarks/deploy-cpu-offload.yaml#L98)

**Comment by amitkumar307d** [comment]
> To enable KV cache offloading, please set TPU_OFFLOAD_NUM_CPU_CHUNKS and TPU_OFFLOAD_NUM_STAGING_BLOCKS. Note that the cpu_bytes_to_use parameter is ignored in tpu-inference.
> 
> Ref: https://github.com/vllm-project/tpu-inference/blob/main/examples/offload/gke/benchmarks/deploy-cpu-offload.yaml#L98

Correct. However, these env vars have default values if not set explicitly, refer https://github.com/vllm-project/tpu-inference/blob/5e0243d1546ed61397d531f070abf3b0b96073a6/tpu_inference/envs.py#L50
